### PR TITLE
docs for stateful handler

### PR DESCRIPTION
**Pull request by patriknw**
References #79

### CODE REVIEW

File: docs/src/main/paradox/cassandra.md [hunk 1]
```diff
@@ -83,7 +83,64 @@ Java
...
+Note that the `state` must be wrapped in a @scala[`Future`]@java[`CompletionStage`] if the handler is performing
+asynchronous operations such as the `repository.loadAll` and `repository.save` in this example. Otherwise it would
+be a high risk that the implementation will have concurrency issues.
+
+Another implementation would be a handler that is loading the current count for a word on demand, and thereafter
+caches it in the in-memory state:
+
+Scala
+:  @@snip [WordCountDocExample.scala](/examples/src/test/scala/docs/cassandra/WordCountDocExample.scala) { #loadingOnDemand }
+
```
Comment (patriknw) [inline docs/src/main/paradox/cassandra.md:143]: I have some ideas of how to alternatively use an actor for more advanced state management. Will be separate PR.

File: examples/src/test/scala/akka/projection/cassandra/scaladsl/WordCountDocExampleSpec.scala [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+      _ <- session.setSchemaMetadataEnabled(null).toScala
+    } yield Done, 30.seconds)
+  }
+
+  override protected def afterAll(): Unit = {
+    // reason for setSchemaMetadataEnabled is that it speed up tests
+    Try(session.underlying().map(_.setSchemaMetadataEnabled(false)).futureValue)
+    Try(session.executeDDL(s"DROP keyspace ${offsetStore.keyspace}").futureValue)
+    Try(session.executeDDL(s"DROP keyspace ${repository.keyspace}").futureValue)
+    Try(session.underlying().map(_.setSchemaMetadataEnabled(null)).futureValue)
```
Comment (patriknw) [inline examples/src/test/scala/akka/projection/cassandra/scaladsl/WordCountDocExampleSpec.scala:60]: note to self: hm, copied from another test, I thought I changed this to for comprehension

File: docs/src/main/paradox/cassandra.md [hunk 3]
```diff
@@ -83,7 +83,64 @@ Java
...
-TODO stateful handlers...
+### Stateful handler
+
+The `Handler` can be stateful, with variables and mutable data structures. It is invoked by the `Projection` machinery
+one envelope at a time and visibility guarantees between the invocations are handled automatically, i.e. no volatile
+or other concurrency primitives are needed for managing the state.
+
+The returned @scala[`Future[Done]`]@java[`CompletionStage<Done>`] is to be completed when then processing of the
+`envelope` has finished. The handler will not be invoked with next envelope until after the returned 
+@scala[`Future[Done]`]@java[`CompletionStage<Done>`] has been completed.
```
Comment (renatocaval) [inline docs/src/main/paradox/cassandra.md:94]:
```suggestion
The returned @scala[`Future[Done]`]@java[`CompletionStage<Done>`] is to be completed when the processing of the
`envelope` has finished. The handler will not be invoked with the next envelope until after the returned 
@scala[`Future[Done]`]@java[`CompletionStage<Done>`] has been completed.
```

File: docs/src/main/paradox/cassandra.md [hunk 4]
```diff
@@ -83,7 +83,64 @@ Java
...
+
+The `Projection` can be definined as:
+
+Scala
+:  @@snip [WordCountDocExample.scala](/examples/src/test/scala/akka/projection/cassandra/scaladsl/WordCountDocExampleSpec.scala) { #projection }
+
+The `handler` can be implemented as follows.
+
+A naive approach would be to have one row per word for storing the current count in the database. 
+The handler could be implemented as a completely stateless handler that for each processed envelope loaded the current
```
Comment (renatocaval) [inline docs/src/main/paradox/cassandra.md:119]:
```suggestion
The handler could be implemented as a completely stateless handler that for each processed envelope loads the current
```

File: examples/src/test/scala/docs/cassandra/WordCountDocExample.scala [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+  }
+  //#sourceProvider
+
+  object IllstrateStatefulHandlerLoadingInitialState {
+
+    //#loadingInitialState
+    class WordCountHandler(projectionId: ProjectionId, repository: WordCountRepository)(implicit ec: ExecutionContext)
+        extends Handler[WordEnvelope] {
+
+      private var state: Future[Map[Word, Count]] = repository.loadAll(projectionId.id)
```
Comment (johanandren) [inline examples/src/test/scala/docs/cassandra/WordCountDocExample.scala:98]: The `Handler` is executed inside of the actor so it is safe to mutate this var without synchronization?

Comment (patriknw) [reply examples/src/test/scala/docs/cassandra/WordCountDocExample.scala:98]: It's executed inside `mapAsync` so it's safe, but wouldn't be good to update such var from any of the Future callbacks. That is what I mean with that the state must be wrapped by a Future.

Comment (johanandren) [reply examples/src/test/scala/docs/cassandra/WordCountDocExample.scala:98]: Feels a bit like a trap waiting to spring on a user

**Comment by renatocaval** [comment]
sorry, closed by accident

File: akka-projection-core/src/main/scala/akka/projection/javadsl/Handler.scala [hunk 6]
```diff
@@ -33,5 +33,12 @@ object Handler {
  */
 @ApiMayChange
 abstract class Handler[Envelope] extends HandlerRecovery[Envelope] {
+
+  /**
+   * The `process` method is invoked for each `Envelope`.
+   * One envelope is processed at a time. The returned `CompletionStage` is to be completed when then processing
+   * of the `envelope` has finished. It will not be invoked with next envelope until after the returned
```
Comment (renatocaval) [inline akka-projection-core/src/main/scala/akka/projection/javadsl/Handler.scala:40]:
```suggestion
   * One envelope is processed at a time. The returned `CompletionStage` is to be completed when the processing
   * of the `envelope` has finished. It will not be invoked with the next envelope until after the returned
```

File: akka-projection-core/src/main/scala/akka/projection/scaladsl/Handler.scala [hunk 7]
```diff
@@ -33,5 +33,13 @@ import akka.projection.HandlerRecovery
  * overriding [[HandlerRecovery.onFailure]].
  */
 @ApiMayChange trait Handler[Envelope] extends HandlerRecovery[Envelope] {
+
+  /**
+   * The `process` method is invoked for each `Envelope`.
+   * One envelope is processed at a time. The returned `Future` is to be completed when then processing
+   * of the `envelope` has finished. It will not be invoked with next envelope until after the returned
```
Comment (renatocaval) [inline akka-projection-core/src/main/scala/akka/projection/scaladsl/Handler.scala:40]:
```suggestion
   * One envelope is processed at a time. The returned `Future` is to be completed when the processing
   * of the `envelope` has finished. It will not be invoked with the next envelope until after the returned
```

File: examples/src/test/scala/docs/cassandra/WordCountDocExample.scala (hunk 5 above)
Comment (patriknw) [reply examples/src/test/scala/docs/cassandra/WordCountDocExample.scala:98]: yes, that's why I started looking at the actor alternative

**Comment by renatocaval** [comment]
> Looks good. I wonder if there is a more functional API we could use instead. Something like the `fold` or `scan` stream operators where the user defines a seed state value and the I/O is stateful object.
> 
> i.e.
> 
> ```
> case class MyState(count: Int)
> 
> class MyHandler extends StatefulHandler[MyState, WordEnvelope] {
> 
>   override val initialState: MyState = MyState(0)
> 
>   override def process(state: MyState, env: WordEnvelope): MyState = {
>     // do something with the envelope if we want..
>     // update state
>     state.copy(count = oldState.count + 1)
>   }
> }
> ```

This is looking like the EventSourcedBehavior/Lagom API. I like it.

**Comment by patriknw** [comment]
> I wonder if there is a more functional API we could use instead. Something like the fold or scan

I'm in general skeptical to that we add more programming models for this than what we already have (actors and streams), but seams simple enough and makes the handler more safe. I'll give it a try in a separate PR.

**Comment by patriknw** [comment]
Travis suddenly doesn't like the Kafka testcontainer. I'll close and open, to see if that helps.
